react-native windows command fails because there is no react-native-windows version compatible with the latest RN

### Repro Steps

I followed the instructions for Getting Started in the README. After running `react-native windows`, I got this error:

```
Installing react-native-windows@0.35.*...
npm ERR! Windows_NT 10.0.10586
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "--save" "react-native-windows@0.35.*"
npm ERR! node v6.5.0
npm ERR! npm  v3.10.3

npm ERR! No compatible version found: react-native-windows@0.35.*
npm ERR! Valid install targets:
npm ERR! 1.0.0, 0.35.0-rc.0, 0.34.1, 0.34.0, 0.34.0-rc.0, 0.33.4, 0.33.3, 0.33.1, 0.33.0-rc.1, 0.33.0-rc.0, 0.32.1, 0.32.0, 0.32.0-rc.0, 0.31.3, 0.31.2, 0.31.1, 0.31.0, 0.31.0-rc.1, 0.31.0-rc.0, 0.30.0, 0.30.0-rc.0, 0.29.2, 0.29.1, 0.29.0, 0.29.0-rc.0, 0.28.3, 0.28.2, 0.28.1, 0.28.0, 0.28.0-rc.1, 0.28.0-rc.0, 0.27.0, 0.27.0-rc, 0.2.5-rc7, 0.2.4, 0.2.3, 0.2.2, 0.2.1, 0.1.0
```

### Cause

I have `react-native` version `0.35.0` installed. I guess this is erroring because it's trying to install `react-native-windows` version `0.35.*` but this version pattern isn't compatible with any of the available `react-native-windows` versions based on the rules of semantic versioning.

### Work Around

Developers can work around this issue by running this command:

```
react-native windows --windowsVersion 0.35.0-rc.0
```

I still haven't published react-native-windows@0.35 yet.

On Oct 12, 2016, at 10:47 PM, Adam Comella <notifications@github.com<mailto:notifications@github.com>> wrote:


I followed the instructions for Getting Started in the README. After running react-native windows, I got this error:

Installing react-native-windows@0.35.*...
npm ERR! Windows_NT 10.0.10586
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "--save" "react-native-windows@0.35.*"
npm ERR! node v6.5.0
npm ERR! npm  v3.10.3

npm ERR! No compatible version found: react-native-windows@0.35.*
npm ERR! Valid install targets:
npm ERR! 1.0.0, 0.35.0-rc.0, 0.34.1, 0.34.0, 0.34.0-rc.0, 0.33.4, 0.33.3, 0.33.1, 0.33.0-rc.1, 0.33.0-rc.0, 0.32.1, 0.32.0, 0.32.0-rc.0, 0.31.3, 0.31.2, 0.31.1, 0.31.0, 0.31.0-rc.1, 0.31.0-rc.0, 0.30.0, 0.30.0-rc.0, 0.29.2, 0.29.1, 0.29.0, 0.29.0-rc.0, 0.28.3, 0.28.2, 0.28.1, 0.28.0, 0.28.0-rc.1, 0.28.0-rc.0, 0.27.0, 0.27.0-rc, 0.2.5-rc7, 0.2.4, 0.2.3, 0.2.2, 0.2.1, 0.1.0


I have react-native version 0.35.0 installed. I guess this is erroring because it's trying to install react-native-windows version 0.35.* but this version pattern isn't compatible with any of the available react-native-windows versions based on the rules of semantic versioning.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/ReactWindows/react-native-windows/issues/778>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABDhPzsjHRrWGaUZhJ8CUEU3pdFVP_1vks5qzZuCgaJpZM4KVbYp>.

--- a/local-cli/rnpm/windows/package.json
+++ b/local-cli/rnpm/windows/package.json
@@ -1,6 +1,6 @@
 {
   "name": "rnpm-plugin-windows",
-  "version": "0.2.0",
+  "version": "0.2.1",
   "description": "rnpm plugin that generates a Windows template project",
   "main": "index.js",
   "scripts": {
@@ -25,6 +25,7 @@
   },
   "dependencies": {
     "semver": "^5.0.3",
-    "chalk": "^1.1.3"
+    "chalk": "^1.1.3",
+    "node-fetch": "^1.3.3"
   }
 }

--- a/local-cli/rnpm/windows/src/windows.js
+++ b/local-cli/rnpm/windows/src/windows.js
@@ -14,6 +14,7 @@ const execSync = require('child_process').execSync;
 const fs = require('fs');
 const path = require('path');
 const semver = require('semver');
+const fetch = require('node-fetch');
 
 const REACT_NATIVE_WINDOWS_GENERATE_PATH = function() {
   return path.resolve(
@@ -34,6 +35,27 @@ const REACT_NATIVE_PACKAGE_JSON_PATH = function() {
   );
 };
 
+function getLatestPackage() {
+  return fetch('https://registry.npmjs.org/react-native-windows?version=latest')
+    .then(result => result && result.ok && result.json());
+}
+
+function checkPackageExists(version) {
+  return fetch(`https://registry.npmjs.org/react-native-windows?version=${version}`)
+    .then(result =>
+    {
+      if (result && result.ok) {
+        return result.json().then(pkg => pkg.version);
+      } else {
+        return getLatestPackage().then(latest => {
+          throw new Error(`Could not find react-native-windows@${version}. ` +
+            `Latest version of react-native-windows is ${latest.version}, try switching to ` +
+            `react-native@${semver.major(latest.version)}.${semver.minor(latest.version)}.*.`);
+        });          
+      }
+    });
+}
+
 function getInstallPackage(version) {
   let packageToInstall = 'react-native-windows';
 
@@ -47,15 +69,13 @@ function getInstallPackage(version) {
     process.exit(1);
   }
 
-  if (validVersion) {
-    packageToInstall += '@' + validVersion;
-  } else if (validRange) {
-    packageToInstall += '@' + version;
-  } else if (version) {
-    // for tar.gz or alternative paths
-    packageToInstall = version;
+  const validVersionOrRange = validVersion || validRange;
+  if (validVersionOrRange) {
+    return checkPackageExists(validVersionOrRange)
+      .then(resultVersion => packageToInstall + '@' + resultVersion);
+  } else {
+    return Promise.resolve(version);
   }
-  return packageToInstall;
 }
 
 function getReactNativeVersion() {
@@ -70,11 +90,15 @@ module.exports = function windows(config, args, options) {
   const ns = options.namespace ? options.namespace : name;
   const version = options.windowsVersion ? options.windowsVersion : getReactNativeVersion();
 
-  const rnwPackage = getInstallPackage(version);
-  console.log(chalk.green(`Installing ${rnwPackage}...`));
-  execSync(`npm install --save ${rnwPackage}`);
-  console.log(chalk.green(`${rnwPackage} successfully installed.`));
+  return getInstallPackage(version)
+    .then(rnwPackage =>
+    {
+      console.log(chalk.green(`Installing ${rnwPackage}...`));
+      execSync(`npm install --save ${rnwPackage}`);
+      console.log(chalk.green(`${rnwPackage} successfully installed.`));
 
-  const generateWindows = require(REACT_NATIVE_WINDOWS_GENERATE_PATH());
-  generateWindows(process.cwd(), name, ns);
-};
+      const generateWindows = require(REACT_NATIVE_WINDOWS_GENERATE_PATH());
+      generateWindows(process.cwd(), name, ns);
+    })
+    .catch(error => console.error(error.message));
+}

--- a/local-cli/rnpm/windows/src/windows.js
+++ b/local-cli/rnpm/windows/src/windows.js
@@ -35,9 +35,10 @@ const REACT_NATIVE_PACKAGE_JSON_PATH = function() {
   );
 };
 
-function getLatestPackage() {
+function getLatestVersion() {
   return fetch('https://registry.npmjs.org/react-native-windows?version=latest')
-    .then(result => result && result.ok && result.json());
+    .then(result => result && result.ok && result.json())
+    .then(result => result.version)
 }
 
 function checkPackageExists(version) {
@@ -47,10 +48,10 @@ function checkPackageExists(version) {
       if (result && result.ok) {
         return result.json().then(pkg => pkg.version);
       } else {
-        return getLatestPackage().then(latest => {
+        return getLatestPackage().then(latestVersion => {
           throw new Error(`Could not find react-native-windows@${version}. ` +
-            `Latest version of react-native-windows is ${latest.version}, try switching to ` +
-            `react-native@${semver.major(latest.version)}.${semver.minor(latest.version)}.*.`);
+            `Latest version of react-native-windows is ${latestVersion}, try switching to ` +
+            `react-native@${semver.major(latestVersion)}.${semver.minor(latestVersion)}.*.`);
         });          
       }
     });
@@ -91,14 +92,12 @@ module.exports = function windows(config, args, options) {
   const version = options.windowsVersion ? options.windowsVersion : getReactNativeVersion();
 
   return getInstallPackage(version)
-    .then(rnwPackage =>
-    {
+    .then(rnwPackage => {
       console.log(chalk.green(`Installing ${rnwPackage}...`));
       execSync(`npm install --save ${rnwPackage}`);
       console.log(chalk.green(`${rnwPackage} successfully installed.`));
 
       const generateWindows = require(REACT_NATIVE_WINDOWS_GENERATE_PATH());
       generateWindows(process.cwd(), name, ns);
-    })
-    .catch(error => console.error(error.message));
+    }).catch(error => console.error(error.message));
 }

